docs/dev/Pagination-v2.md

```diff
@@ -100,10 +111,381 @@ The discussion below uses *under max_result_window* to refer to scenarios that c
+    state "Parse Tree" as Parse
+    state "Unresolved Plan Tree" as Unresolved
+    state "Abstract Plan Tree" as Abstract
+    state "Logical Plan Tree" as Logical
+    state "Optimized Logical Plan Tree" as Optimized
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Parse : ANTLR
+    Parse --> Unresolved : AstBuilder
+    Unresolved --> Abstract : QueryPlanner
+    Abstract --> Logical : Planner
+    Logical --> Optimized : Optimizer
+    Optimized --> Physical : Implementor
+  }
+```
+```mermaid
+stateDiagram-v2
+  state "Initial Page Request" as Paged {
+    direction LR
+    state "Parse Tree" as Parse
+    state "Unresolved Plan Tree" as Unresolved
+    state "Abstract Plan Tree" as Abstract
+    state "Logical Plan Tree" as Logical
+    state "Optimized Logical Plan Tree" as Optimized
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Parse : ANTLR
+    Parse --> Unresolved : AstBuilder
+    Unresolved --> Abstract : QueryPlanner
+    Abstract --> Logical : Planner
+    Logical --> Optimized : Optimizer
+    Optimized --> Physical : Implementor
+  }
+```
+```mermaid
+stateDiagram-v2
+  state "Subsequent Page Request" as Paged {
+    direction LR
+    state "Abstract Plan Tree" as Abstract
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Abstract : QueryPlanner
+    Abstract --> Physical : Deserializer
+  }
+```
+
+New plan tree workflow was added for Subsequent Page Requests. Since a final Physical Plan tree was already created for Initial request, subsequent requests should have the same tree. The tree is being serialized into `cursor` by `PlanSerializer` and then restored back. Query parsing and analysis is also not performed for cursor requests, so these steps are also not executed there.
+
+#### Abstract Plan tree
+
+Abstract Plan Tree for non-paged requests remains unchanged by pagination feature. `QueryPlan` as a top entry  got new optional field `pageSize`, which is unset for non-paged requests.
+
+```mermaid
+classDiagram
+  direction LR
+  class QueryPlan {
+    <<AbstractPlan>>
+    -Optional~int~ pageSize
+    -UnresolvedPlan plan
+    -QueryService queryService
+  }
+  class UnresolvedPlanTree {
+    <<UnresolvedPlan>>
+  }
+  QueryPlan --* UnresolvedPlanTree
+```
+
+Abstract plan tree for Initial Query Request has following changes:
+1. New Plan node -- `Paginate` -- added into the tree.
+2. `pageSize` parameter in `QueryPlan` is set, and `Paginate` is being added. It is converted to `LogicalPaginate` later.
+
+```mermaid
+classDiagram
+  direction LR
+  class QueryPlan {
+    <<AbstractPlan>>
+    -Optional~int~ pageSize
+    -UnresolvedPlan plan
+    -QueryService queryService
+  }
+  class Paginate {
+    <<UnresolvedPlan>>
+    -int pageSize
+    -UnresolvedPlan child
+  }
+  class UnresolvedPlanTree {
+    <<UnresolvedPlan>>
+  }
+  QueryPlan --* Paginate
+  Paginate --* UnresolvedPlanTree
+```
+
+Non-paging requests have the same plan tree, but `pageSize` value in `QueryPlan` is unset.
```

Comment:
Is it necessary to say this twice?

Reply:
I can't find the second time, where is it?